lib/Dialect/FIRRTL/Transforms/LowerXMR.cpp

```diff
@@ -208,7 +211,9 @@ class LowerXMRPass : public LowerXMRBase<LowerXMRPass> {
       if (!refMod)
         return inst.emitOpError("cannot lower ext modules with RefType ports");
       // Reference ports must be removed.
-      refPortsToRemoveMap[inst].push_back(portNum);
+      if (refPortsToRemoveMap[inst].size() < portNum + 1)
```

Comment:
I think this should be sized to number of ports on the instance, as it will be passed to `InstanceOp::erasePorts` later, which may test bits for all arguments and expect that to work.

Indeed, tweaking the `lowerXMR.mlir` test case can trigger a crash with this as-is:

```diff
diff --git a/test/Dialect/FIRRTL/lowerXMR.mlir b/test/Dialect/FIRRTL/lowerXMR.mlir
index 7bb9397ac..71e793420 100644
--- a/test/Dialect/FIRRTL/lowerXMR.mlir
+++ b/test/Dialect/FIRRTL/lowerXMR.mlir
@@ -19,16 +19,18 @@ firrtl.circuit "xmr" {
 // Test the correct xmr path is generated
 // CHECK-LABEL: firrtl.circuit "Top" {
 firrtl.circuit "Top" {
-  firrtl.module @XmrSrcMod(out %_a: !firrtl.ref<uint<1>>) {
+  firrtl.module @XmrSrcMod(out %_a: !firrtl.ref<uint<1>>, out %b: !firrtl.uint<1>, out %c : !firrtl.uint<1>) {
     // CHECK: firrtl.module @XmrSrcMod() {
     %zero = firrtl.constant 0 : !firrtl.uint<1>
     // CHECK:  %c0_ui1 = firrtl.constant 0 : !firrtl.uint<1>
     // CHECK:  %0 = firrtl.node sym @xmr_sym %c0_ui1  : !firrtl.uint<1>
     %1 = firrtl.ref.send %zero : !firrtl.uint<1>
     firrtl.strictconnect %_a, %1 : !firrtl.ref<uint<1>>
+    firrtl.strictconnect %b, %zero : !firrtl.uint<1>
+    firrtl.strictconnect %c, %zero : !firrtl.uint<1>
   }
   firrtl.module @Bar(out %_a: !firrtl.ref<uint<1>>) {
-    %xmr   = firrtl.instance bar sym @barXMR @XmrSrcMod(out _a: !firrtl.ref<uint<1>>)
+    %xmr, %xmrb, %xmrc   = firrtl.instance bar sym @barXMR @XmrSrcMod(out _a: !firrtl.ref<uint<1>>, out b: !firrtl.uint<1>, out c: !firrtl.uint<1>)
     // CHECK:  firrtl.instance bar sym @barXMR  @XmrSrcMod()
     firrtl.strictconnect %_a, %xmr   : !firrtl.ref<uint<1>>
   }
```

I think here too we should only add an entry if there are reftype ports to remove (as it does presently).

Reply:
Thanks for catching this! I'm now resizing it to the maximum number of ports in https://github.com/llvm/circt/pull/3837/commits/b278e15b1cc4437d64118c72e53802bef4d1c732